rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.h

```diff
@@ -514,6 +515,12 @@ void deserialize_array(
     (void)call_new;
     if (member->array_size_ && !member->is_upper_bound_) {
         deser.deserializeArray((T*)field, member->array_size_);
+    } else {
+        auto & data = *reinterpret_cast<typename GenericCArray<T>::type *>(field);
+        int32_t dsize;
+        deser >> dsize;
+        GenericCArray<T>::init(&data, dsize);
```

Comment:
Since `field` is a  ROS 2 array the corresponding `Array__init` function in `rosidl_generator_c` must be used.